connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Loggers.java

```diff
@@ -52,184 +53,274 @@ public class Loggers {
-                    break;
-                }
-            }
-        }
-
-        if (foundLogger == null) {
-            log.warn("Unable to find level for logger {}", logger);
-            return null;
-        }
-
-        return loggerLevel(foundLogger);
-    }
+    public abstract LoggerLevel level(String loggerName);
 
     /**
      * Retrieve the current levels of all known loggers
+     *
      * @return the levels of all known loggers; may be empty, but never null
      */
-    public synchronized Map<String, LoggerLevel> allLevels() {
-        return currentLoggers()
-            .values()
-            .stream()
-            .filter(logger -> !logger.getLevel().equals(Level.OFF))
-            .collect(Collectors.toMap(
-                this::getLoggerName,
-                this::loggerLevel,
-                (existing, replacing) -> replacing,
-                TreeMap::new)
-            );
-    }
+    public abstract Map<String, LoggerLevel> allLevels();
 
     /**
      * Set the level for the specified logger and all of its children
+     *
      * @param namespace the name of the logger to adjust along with its children; may not be null
-     * @param level the level to set for the logger and its children; may not be null
+     * @param level     the level to set for the logger and its children; may not be null
      * @return all loggers that were affected by this action, sorted by their natural ordering;
      * may be empty, but never null
      */
-    public synchronized List<String> setLevel(String namespace, Level level) {
-        Objects.requireNonNull(namespace, "Logging namespace may not be null");
-        Objects.requireNonNull(level, "Level may not be null");
-        String internalNameSpace = isValidRootLoggerName(namespace) ? LogManager.ROOT_LOGGER_NAME : namespace;
-
-        log.info("Setting level of namespace {} and children to {}", internalNameSpace, level);
-
-        var loggers = loggers(internalNameSpace);
-        var nameToLevel = allLevels();
-
-        List<String> result = new ArrayList<>();
-        Configurator.setAllLevels(internalNameSpace, level);
-        for (org.apache.logging.log4j.Logger logger : loggers) {
-            // We need to track level changes for each logger and record their update timestamps  to ensure this method
-            // correctly returns only the loggers whose levels were actually modified.
-            String name = getLoggerName(logger);
-            String newLevel = logger.getLevel().name();
-            String oldLevel = nameToLevel.getOrDefault(name, new LoggerLevel("", time.milliseconds())).level();
-            if (!newLevel.equalsIgnoreCase(oldLevel)) {
-                lastModifiedTimes.put(name, time.milliseconds());
-                result.add(name);
+    public abstract List<String> setLevel(String namespace, String level);
+
+    public abstract boolean isValidLevel(String level);
+
+    static class Log4jLoggers extends Loggers {
```

Comment:
Maybe we can move it to a separate file and then limit the log4j import via `checkstyle`? That can ensure we don't use log4j code in other place, such as `LoggingResource`

Reply:
In the long term I would use [apache/logging-admin](https://github.com/apache/logging-admin) (as soon as we publish it), which abstracts the way loggers are configured and provides a limited access to the capabilities of the logging implementations.

For now the main goal is not to expose any of these details (classes, methods, etc.) as public API, so that they could be removed in a minor version of Kafka 4.x, without breaking changes.

Checkstyle is not bulletproof, but it should be enough for now. Fixed in https://github.com/apache/kafka/pull/18496/commits/db05e49b49836e9392aaf1f3c0d6a27124ec4445